akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -247,13 +260,24 @@ object StreamLayout {
       val matComputation1 = if (this.isSealed) Atomic(this) else this.materializedValueComputation
       val matComputation2 = if (that.isSealed) Atomic(that) else that.materializedValueComputation
 
+      val mat =
+        {
+          val comp =
+            if (f == scaladsl.Keep.left) {
+              if (IgnorableMatValComp(matComputation2)) matComputation1 else null
+            } else if (f == scaladsl.Keep.right) {
+              if (IgnorableMatValComp(matComputation1)) matComputation2 else null
+            } else null
+          if (comp == null) Combine(f.asInstanceOf[(Any, Any) ⇒ Any], matComputation1, matComputation2)
+          else comp
+        }
```

Comment:
this function will not help the user defined `Keep.left` & `Keep.right`,maybe should metion it on doc.like we do special optimize when you use `Keep.left` & `Keep.right`,so you should not write your own ones.